### PR TITLE
Check SPI attribute usage in frozen types and on protocol requirements

### PR DESCRIPTION
**Pull request by xymus**
Add two soundness checks on the use of the `@_spi` attribute:
* Report stored properties marked with `@_spi` in frozen structs as an error. A frozen struct with an SPI stored properties would prevent clients to know the correct layout of the struct.
* Check that SPI requirements in public protocols have a default implementation defined in an unconstrained protocol extension in the same module.

### CODE REVIEW

File: lib/Sema/TypeCheckAttr.cpp [hunk 1]
```diff
@@ -863,13 +863,61 @@ void AttributeChecker::visitSetterAccessAttr(
...
+      bool hasDefaultImplementation = llvm::any_of(implementations,
+        [&](const LookupResultEntry &entry) {
+          auto DC = entry.getValueDecl()->getDeclContext();
+          auto extension = dyn_cast<ExtensionDecl>(DC);
+
+          // The implementation must be defined in the same module in
+          // an unconstrained extension.
+          if (!extension ||
+              extension->getParentModule() != protocol->getParentModule() ||
+              extension->isConstrainedExtension()) return false;
```
Comment (nkcsgexi) [inline lib/Sema/TypeCheckAttr.cpp:892]: Nitpick: move `return false;` to the next line for better debuggability.

File: lib/Sema/TypeCheckAttr.cpp [hunk 2]
```diff
@@ -863,13 +863,61 @@ void AttributeChecker::visitSetterAccessAttr(
...
+
+          if (auto protoStorage = dyn_cast<AbstractStorageDecl>(VD))
```
Comment (nkcsgexi) [inline lib/Sema/TypeCheckAttr.cpp:894]: Nitpick: could you add some comments when the special handling of `AbstractStorageDecl` is needed here?

**Comment by xymus** [comment]
@swift-ci please smoke test

**Review by nkcsgexi — COMMENTED** [review]
Thank you fixing these!

File: lib/Sema/TypeCheckAttr.cpp (hunk 2 above)
Comment (xymus) [reply lib/Sema/TypeCheckAttr.cpp:898]: Done! I also simplified the check. We only need a special check for the definition of `set` as `get` is already required.

**Comment by xymus** [comment]
@swift-ci please smoke test

**Review by beccadax — APPROVED** [review]

**Pull request merged**